Arch arm secure fault handler implementation cm23 and rework

ioannisg opened the pull request:
This PR finalizes the implementation of Secure Fault handling for ARM CPUs that implement the ARMv8-m Security Extension.
- Contributes the implementation for CM23
- Reworks the fault dumping function to give more precise information to the user regarding what has occurred during a secure fault.

Addresses #6372

ioannisg committed:
arch: arm: Secure fault handling for Cortex-M23

This commit contributes the Secure fault handling for ARMv8-m
implementations, that do not support the Main Extension, such
as the (Baseline) Cortex-M23 processor.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg committed:
arch: arm: improve fault dump for secure firmware

This commit partially reworks the fault dump implementation
of faults in non-secure state that target Secure state. In
particular, the return address to secure state is logged along
with the non-secure address which caused the non-secure fault
(that targets the Secure state).

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg committed:
arch: arm: distinguish integrity signatures with/without FP

Distinguish between the integrity signatures depending on whether the
ARM MCU implements the Floating-Point Extension or not.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg commented:
Added extra commit (originally planned for following PR), since review of this PR has not started yet.

carlescufi commented on arch/arm/core/fault_s.S:
```diff
@@ -109,10 +110,35 @@ _stack_frame_endif:
 				 *  this is not a nested exception: the stack
 				 * frame is on the PSP */
+#else
+	/* RETTOBASE flag is not bancked between security states.
```
banked

ioannisg replied:
Will fix.

carlescufi commented on arch/arm/core/fault_s.S:
```diff
@@ -109,10 +110,35 @@ _stack_frame_endif:
 #endif /* CONFIG_ARMV6_M_ARMV8_M_BASELINE */
 
+#if defined(CONFIG_ARM_SECURE_FIRMWARE)
+	/* In an ARM Secure firmware, the stack pointer that is retrieved
```
remove "an"

ioannisg replied:
Will fix.

carlescufi commented on arch/arm/core/fault.c:
```diff
@@ -37,6 +37,69 @@
+/* Prefix. Indicates that this is an EXC_RETURN value.
+ * This field reads as 0b11111111.
+ */
+#define EXC_RETURN_INDICATOR_PREFIX     (0xFF << 24)
```
Doesn't CMSIS contain anything about the `EXC_RETURN` mappings?

ioannisg replied:
No, it does not.

galak replied:
Do you mind filing an issue on the CMSIS_5 project

ioannisg replied:
https://github.com/ARM-software/CMSIS_5/issues/340

for both EXC_RETURN and the Integrity Signature!

ioannisg commented:
@carlescufi Re-uploaded with fixed changes.

ioannisg commented:
@galak any feedback/comments on this one?

galak commented on arch/arm/core/fault.c:
```diff
@@ -490,11 +540,87 @@ static void _FaultDump(const NANO_ESF *esf, int fault)
  *            state the exception was taken.
  */
+#if defined(CONFIG_ARM_SECURE_FIRMWARE)
+void _Fault(const NANO_ESF *esf, u32_t exc_return)
```
can we have the same function signature in both cases and just mark exc_return usused in the CONFIG_ARM_SECURE_FIRMWARE case.

ioannisg replied:
Alright, I am going to modify this accordingly.

galak commented on arch/arm/core/fault.c:
```diff
@@ -37,6 +37,69 @@
+#define EXC_RETURN_RETURN_STACK_Secure EXC_RETURN_RETURN_STACK_Msk
+
+/* Integrity signature for an ARMv8-M implementation */
+#if !defined(CONFIG_ARMV7_M_ARMV8_M_FP)
```
change this from if !defined() to if defined()

ioannisg replied:
sure

galak reviewed (changes requested):
a few changes requested.

ioannisg commented:
@galak implemented changes & rebased

ioannisg committed:
arch: arm: fix bug in AIRCR config on init

Fix bug when setting SCB_AIRCR during initialization; add
Vector Key to permit write to AIRCR fields.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg commented:
- Re-based once more.
- Added a separate commit to fix a bug when overwriting the AIRCR register.

ioannisg committed:
arch: arm: fix SecureFault IRQn for non-CMSIS compliant MCUs

This commit fixes the SecureFault IRQ number for non-CMSIS
compliant ARM Cortex-M MCUs. Erroneous IRQ number was introduced
in ec398e87796f8da2529f369746ae9b2b12d48ca4.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

carlescufi commented:
@galak can you take another look please?
